### PR TITLE
Support load-balanced TPU Tokamax ring attention

### PR DESCRIPTION
**Pull request by huytransformer**
# Description

This PR adds load-balancing support (`context_parallel_load_balance=true`) to TPU Tokamax ring attention (`context_parallel_strategy=ring` + `use_tokamax_splash=true`), following #4266 which added the unbalanced ring.

**Why.** With causal masking and normal sequence sharding, shard 0 attends to almost nothing while the last shard attends to nearly the full sequence. At high CP, roughly half of attention compute is wasted.

**How.** Reuses MaxText current `DUAL_CHUNK_SWAP`  and makes the splash kernel mask aware of the permutation:
- A `kv_sequence` array of original token positions travels with each rotating KV chunk.
- `CausalMask`/`MaskInfo` compare original Q positions against original KV positions, so masking is exact under the permutation.

# Key Changes

- `kernels/tokamax_splash_attention/splash_attention_mask_info.py`: mask info is built from `kv_sequence`, the original token positions of the permuted KV.
- `kernels/tokamax_splash_attention/splash_attention_kernel.py`: kernels compare original Q positions against original KV positions instead of physical block indices.
- `kernels/attention/tokamax_ring_attention.py`: each ring hop passes the `kv_sequence` slice along with its KV chunk.
- `layers/attention_op.py`: removes the ring rejection of `context_parallel_load_balance`, so the existing DUAL_CHUNK_SWAP reorder now applies to ring.
- `configs/types.py`: rejects unsupported configs at startup: odd `ici_context_parallelism`, MTP with LB.

# Tests

llama3.1-8b on v5p 64 chips 

**TLDR:** load balancing gives **1.34× / 1.45× / 1.37×** end-to-end training speedup at 256K / 1M / 64K context with same loss curves.

- Tokamax ring attention + mask unit tests: all **13 passed** on TPU.
- Config validation tests (`configs_value_test -k tpu_tokamax_ring`): all **3 passed**.

# Performance Impact

Setup: llama3.1-8b, v5p-128, synthetic data, bf16, `sa_block_*=1024`, `dq_reduction_steps=3`, `remat_policy=full`, `num_vocab_tiling=16`.

Seq len | Mesh (cp×fsdp) | LB | Step time (s) | TFLOP/s/device | MFU (v5p, 459 peak) | Tokens/s/device | Speedup
-- | -- | -- | -- | -- | -- | -- | --
64K | 4×16 | off | 18.14 | 174.4 | 38.0% | 1806 | 1.00×
64K | 4×16 | on | 13.92 | 227.3 | 49.5% | 2354 | 1.30×
256K | 16×4 | off | 63.32 | 130.0 | 28.3% | 517 | 1.00×
256K | 16×4 | on | 41.24 | 199.6 | 43.5% | 795 | 1.54×
1M | 64×1 | off | 242.9 | 117.3 | 25.6% | 135 | 1.00×
1M | 64×1 | on | 150.3 | 189.6 | 41.3% | 218 | 1.62×

Memory: identical memory use.

Loss equivalence: over 100 steps at 64K near identical loss (same up to 8e-3).   

**Xprofs:**
- 256K off: [xprof](https://xprof.corp.google.com/overview_page/htn-12969162626154064340)
- 256K on: [xprof](https://xprof.corp.google.com/overview_page/htn-6255156894659535418)
- 1M off: [xprof](http://xprof.corp.google.com/trace_viewer/htn-3914347130787873298)
- 1M on: [xprof](http://xprof.corp.google.com/trace_viewer/htn-4512027334587333121)

Repro (256K uses `max_target_length=262144 ici_context_parallelism=16 ici_fsdp_parallelism=4 per_device_batch_size=0.125`):

```bash
python3 -m maxtext.trainers.pre_train.train src/maxtext/configs/base.yml \
  base_output_directory=$GCS dataset_type=synthetic enable_checkpointing=false \
  model_name=llama3.1-8b num_vocab_tiling=16 \
  attention=flash use_tokamax_splash=true context_parallel_strategy=ring packing=false \
  sa_block_q=1024 sa_block_kv=1024 sa_block_q_dkv=1024 sa_block_kv_dkv=1024 \
  dq_reduction_steps=3 remat_policy=full \
  max_target_length=1048576 ici_context_parallelism=64 ici_fsdp_parallelism=1 \
  per_device_batch_size=0.03125 steps=10 \
  context_parallel_load_balance=true run_name=$RUN
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):

- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: src/maxtext/configs/types.py [hunk 1]
```diff
@@ -3409,7 +3409,10 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
       if self.packing:
         raise ValueError("TPU Tokamax ring attention does not support packing yet.")
       if self.context_parallel_load_balance:
-        raise ValueError("TPU Tokamax ring attention does not support context_parallel_load_balance yet.")
+        if context_parallel_size % 2 != 0:
+          raise ValueError("TPU Tokamax ring load balancing requires an even context_parallel_size.")
+        if self.mtp_num_layers > 0:
+          raise ValueError("TPU Tokamax ring attention with context_parallel_load_balance=True does not support MTP.")
```
Comment (NuojCheng) [inline src/maxtext/configs/types.py:3415]: why mtp not supported?

File: src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py [hunk 2]
```diff
@@ -207,30 +208,38 @@ def _apply_mask_and_soft_cap(
...
       else:
         assert q_sequence_ref.shape == (NUM_SUBLANES, bq)
 
-        k_sequence = k_offset + jax.lax.broadcasted_iota(jnp.int32, (k_slice.size, bq), 0)
+        if kv_sequence_ref is None:
+          k_sequence = k_offset + jax.lax.broadcasted_iota(jnp.int32, (k_slice.size, bq), 0)
+        else:
+          repeats, rem = divmod(bq, NUM_LANES)
+          if rem:
+            raise NotImplementedError(f"block_q must be a multiple of {NUM_LANES}")
```
Comment (NuojCheng) [inline src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py:241]: move this to types.py?

File: src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py [hunk 3]
```diff
@@ -394,6 +404,7 @@ def body(kv_compute_index, _, has_partial_mask=False):
         mask_value,
         mask_ref,
         q_sequence_ref,
+        kv_sequence_ref,
```
Comment (NuojCheng) [inline src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py:407]: I am surprised kv_sequence_ref previously was not part of inputs..

File: src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py (hunk 2 above)
Comment (gobbleturk) [reply src/maxtext/kernels/tokamax_splash_attention/splash_attention_kernel.py:241]: might be hard to plumb logic involving NUM_LANES to types.py, I think this is okay here

**Review by gobbleturk — APPROVED** [review]

File: src/maxtext/configs/types.py (hunk 1 above)
Comment (huytransformer) [reply src/maxtext/configs/types.py:3415]: perhaps it can be lifted later when I know for sure how it works

**Review by NuojCheng — APPROVED** [review]
